src/rust/inetstack/protocols/tcp/queue.rs

```diff
@@ -43,8 +49,11 @@ impl<const N: usize> TcpQueue<N> {
     }
 
     /// Set underlying TCP socket data structure.
-    pub fn set_socket(&mut self, s: Socket<N>) {
-        self.socket = s;
+    pub fn set_socket(&mut self, mut s: Socket<N>) {
+        swap(&mut s, &mut self.socket);
+
+        // Don't run the Socket destructor at this moment.
+        forget(s);
```

Comment:
Instead of doing this, maybe we should more explicitly remove the Task from the scheduler, instead of in the drop function? It makes it hard to debug when there are implicit drops that manipulate the scheduler because it's hard to find where all references are dropped.

Reply:
Adding more to this... Why do we want to skip the destructor?